### ISSUE TITLE
An example showing significant whitespace?

### ISSUE THREAD

**Issue by Geordi7** [kach/nearley#643]
I'm having difficulty creating a parser for a language like Pug, I haven't tried using an external lexer, but I have a sneaking suspicion it is necessary.

Can you provide an example which shows how to do it?

**Comment by TekuConcept** [comment]
> significant whitespace

As in multiple contiguous whitespace characters?
```
OMS -> [\s]:* # optional multi-line whitespace
RMS -> [\s]:+ # required multi-line whitespace
```

**Comment by Geordi7** [comment]
No, as in scopes delimited by indented sections of text (al a pug python haskell, etc.)

**Comment by TekuConcept** [comment]
> as in scopes delimited by indented sections of text (al a pug python haskell, etc.)

Ah, so indent / dedent... that will be a context-aware parsing solution.

## Use local state
You could get away with creating and updating a local context in the grammar post-processing step, eg.

```
LINES
    -> LINES RBS LINE {% d => {
        // where d[0] is the state object
        d => updateState(d)
    } %}
    |  LINE {% d => createState(d) %}
    
RBS -> OWS LF OMS # required break space
OMS -> [\s]:*     # optional multi-line space
OWS -> [ \t\r]:*  # optional white space
LF -> "\n"
```

This technique, however, will pose a few challenges and limitations, but it's one way to go about this without creating your own lexer.

## Use a custom lexer
This may perhaps be the more trivial way of parsing indent / dedent - as your _sneaking suspicion_ was hinting to. _(Haven't tried it myself yet.)_ I found the following on moo's issue tracker for context-aware indent / dedent parsing: https://github.com/no-context/moo/issues/55 with the last link ([moo-indentation-lexer](https://github.com/aliclark/moo-indentation-lexer)) being the one you probably want.

Then according to the [nearley docs](https://nearley.js.org/docs/tokenizers):
```
@{%
    const moo = require("moo")
    const IndentationLexer = require('moo-indentation-lexer')

    // Create a lexer from rules
    const mooLexer = moo.compile({ ... })
    // Create an indentation-aware lexer using the lexer
    const lexer = new IndentationLexer({ lexer: mooLexer })
%}

# Pass your lexer object using the @lexer option:
@lexer lexer

BLOCK -> HEADING %indent STATEMENTS %dedent
```
